compiler/incremental-compilation-impl/src/org/jetbrains/kotlin/incremental/classpathDiff/ProgramSymbol.kt

```diff
@@ -20,97 +22,172 @@ data class ClassMember(val classId: ClassId, val memberName: String) : ProgramSy
 
 data class PackageMember(val packageFqName: FqName, val memberName: String) : ProgramSymbol()
 
-/** Compact representation for a set of [PackageMember]s. */
-class PackageMemberSet(val packageToMembersMap: Map<FqName, Set<String>>) {
+/** Compact representation for a set of [ProgramSymbol]s. */
+class ProgramSymbolSet private constructor(
 
-    operator fun contains(other: PackageMember): Boolean {
-        return packageToMembersMap[other.packageFqName]?.let { other.memberName in it } ?: false
+    /** Compact set of [ClassSymbol]s. */
+    val classes: Set<ClassId>,
+
+    /** Compact set of [ClassMember]s (map from a [ClassId] to the class members' names). */
+    val classMembers: Map<ClassId, Set<String>>,
+
+    /** Compact set of [PackageMember]s (map from a package's [FqName] to the package members' names). */
+    val packageMembers: Map<FqName, Set<String>>
+) {
+
+    fun isEmpty() = classes.isEmpty() && classMembers.all { it.value.isEmpty() } && packageMembers.all { it.value.isEmpty() }
+
+    operator fun plus(other: ProgramSymbolSet): ProgramSymbolSet {
+        return Collector().run {
+            addClasses(classes)
+            addClasses(other.classes)
+
+            classMembers.forEach { addClassMembers(it.key, it.value) }
+            other.classMembers.forEach { addClassMembers(it.key, it.value) }
+
+            packageMembers.forEach { addPackageMembers(it.key, it.value) }
+            other.packageMembers.forEach { addPackageMembers(it.key, it.value) }
+
+            getResult()
+        }
+    }
+
+    fun asSequence(): Sequence<ProgramSymbol> {
+        return classes.asSequence().map { ClassSymbol(it) } +
+                classMembers.asSequence().flatMap { (classId, memberNames) -> memberNames.asSequence().map { ClassMember(classId, it) } } +
+                packageMembers.asSequence()
+                    .flatMap { (packageFqName, memberNames) -> memberNames.asSequence().map { PackageMember(packageFqName, it) } }
     }
 
-    fun containsElementsIn(other: PackageMemberSet): Boolean {
-        return this.packageToMembersMap.keys.intersect(other.packageToMembersMap.keys).any { commonPackage ->
-            val otherMembers = other.packageToMembersMap[commonPackage]!!
-            this.packageToMembersMap[commonPackage]!!.any { it in otherMembers }
+    inline fun forEach(action: (ProgramSymbol) -> Unit) = asSequence().forEach(action)
```

Comment:
It seems like `forEach` (which also uses `plus` and `asSequence`) is not really needed if `toChangesEither` is moved into `ProgramSymbolSet`? That also avoid creating intermediate symbols, only to immediately convert them to LookupSymbol.

Reply:
Yeah `forEach` is optional, looks like your suggestion is to remove `asSequence`?

`ProgramSymbolSet` is equivalent to `Set<ProgramSymbol>`, so we need an API to iterate through each element. Converting `ProgramSymbolSet` to `Set<ProgramSymbol>` would be a bit memory-inefficient, so we have `ProgramSymbolSet.asSequence` to provide a `Sequence<ProgramSymbol>` instead.

For a user of `ProgramSymbolSet`, it would be nicer to work with `Sequence<ProgramSymbol>` than to work with `ProgramSymbolSet`'s  internal properties.

I can make `ProgramSymbolSet.toChangesEither()` work with `ProgramSymbolSet`'s  internal properties directly if you think it's important for performance, but I'll probably keep that method outside of `ProgramSymbolSet` as that method is bit of a distraction from the main implementation of `ProgramSymbolSet`.

WDYT?

Reply:
Okay I've removed the `forEach` method. I still need the `asSequence` method because:

`ChangesEither.Known` is essentially `Collection<LookupSymbol>`, so here we're converting a collection of `ProgramSymbol`s to a collection of `LookupSymbol`s.

We already have the `ProgramSymbol.toLookupSymbol(): LookupSymbol` method (used in several places) to convert a `ProgramSymbol` to a `LookupSymbol`.

So the missing piece is to get a collection of `ProgramSymbol`s for the input, which this `asSequence` method does.


I've considered all the points you said, but let me know if this looks acceptable now. I'll revisit this if it becomes a performance issue later (showing up during profiling).

Reply:
Thanks, keeping it like this looks good to me.